File: React_Frontend-main/frontend/src/lib/api/logisticsDashboard.ts
import { apiFetch } from './client'
import type { ReferenceSet } from '@/components/ReferenceList'
import type { ResolvedPeriod, Coverage } from '@/components/PeriodFilter'
import type { DataNote } from '@/components/DataNotes'

/** Every tab carries the same three things as the other dashboards: the window
 *  it actually used, what its source holds, and where a figure rests on a
 *  partly-filled column. Logistics was the one screen without them. */
export interface TabMeta {
  period: ResolvedPeriod
  coverage: Coverage
  dataNotes: DataNote[]
  dateField: string
  dateFieldOptions: { value: string; label: string }[]
}

/**
 * Logistics is three endpoints, not one — /dashboard/logistics/{shipments,
 * packing,transport} — matching the three tabs on the page. Each returns
 * finished figures plus its own filter option lists, the same contract as the
 * purchases and inventory dashboards, so the views render them directly.
 *
 * There is no documentation endpoint: that tab was removed for lack of data.
 */

export interface LabelValue {
  [key: string]: unknown
  label: string
  value: number
}

function buildQuery(filters: object): string {
  const params = new URLSearchParams()
  for (const [key, value] of Object.entries(filters)) {
    if (Array.isArray(value)) {
      for (const v of value) params.append(key, String(v))
    } else if (value) {
      params.set(key, String(value))
    }
  }
  const qs = params.toString()
  return qs ? `?${qs}` : ''
}

// ------------------------------------------------------------------ shipments

export interface ShipmentKpis {
  shipments_shown: number
  delivered: number
  not_yet_linked: number
  total_cost: number
  avg_cost_per_kg: number
  countries: number
}

export interface ShipmentsFilters {
  status?: string[]
  stage?: string[]
  shipping_line?: string[]
  country?: string[]
  customer?: string[]
  etd_from?: string
  etd_to?: string
  search?: string
  /** The dashboard-wide window. Omit BOTH for the current month. */
  date_from?: string
  date_to?: string
  /** Which date it applies to: etd (sailing) | eta (arrival). */
  date_field?: string
}

export interface OrderTypeSplit {
  export: number
  local: number
  /** Orders whose type the sheet never states — shown rather than folded
   *  into either side. */
  not_stated: number
  total: number
}

export interface OrderTypeCounts extends OrderTypeSplit {
  windowed: boolean
  /** Orders with NO date in the chosen column, so they fall in no period at
   *  all. NOT ONE local order carries a business date, so without this the
   *  local tile is a silent zero that reads as "we do no local business". */
  undated: OrderTypeSplit
}

export interface ShipmentsResponse extends TabMeta {
  kpis: ShipmentKpis
  references: {
    orders: ReferenceSet
    delivered: ReferenceSet
    not_linked: ReferenceSet
    export: ReferenceSet
    local: ReferenceSet
    not_stated: ReferenceSet
    /** The orders no window reaches. */
    undated: ReferenceSet
  }
  /** Export against local IN THE WINDOW, with the undated orders alongside. */
  orderTypeCounts: OrderTypeCounts
  statusSplit: LabelValue[]
  costPerKgByCountry: LabelValue[]
  statuses: string[]
  stages: string[]
  shippingLines: string[]
  countries: string[]
  customers: string[]
}

export async function getShipmentsDashboard(filters: ShipmentsFilters = {}): Promise<ShipmentsResponse> {
  const { data } = await apiFetch<{ data: {
    kpis: ShipmentKpis
    status_split: LabelValue[]; cost_per_kg_by_country: LabelValue[]
    statuses: string[]; stages: string[]; shipping_lines: string[]
    countries: string[]; customers: string[]
    references: ShipmentsResponse['references']
    order_type_counts: OrderTypeCounts
    period: ResolvedPeriod
    coverage: Coverage
    data_notes: DataNote[]
    date_field: string
    date_field_options: { value: string; label: string }[]
  } }>(`/dashboard/logistics/shipments${buildQuery(filters)}`)
  return {
    kpis: data.kpis,
    statusSplit: data.status_split,
    costPerKgByCountry: data.cost_per_kg_by_country,
    statuses: data.statuses,
    stages: data.stages,
    shippingLines: data.shipping_lines,
    countries: data.countries,
    customers: data.customers,
    references: data.references,
    orderTypeCounts: data.order_type_counts,
    period: data.period,
    coverage: data.coverage,
    dataNotes: data.data_notes,
    dateField: data.date_field,
    dateFieldOptions: data.date_field_options,
  }
}

// -------------------------------------------------------------------- packing

export interface PackingKpis {
  packing_jobs_shown: number
  packed: number
  total_cost: number
  avg_rfd_delay_days: number | null
  categories: number
}

export interface PackingFilters {
  status?: string[]
  works?: string[]
  product_category?: string[]
  business_type?: string[]
  customer?: string[]
  packing_from?: string
  packing_to?: string
  search?: string
  date_from?: string
  date_to?: string
  /** packed | rfd */
  date_field?: string
}

export interface PackingResponse extends TabMeta {
  kpis: PackingKpis
  references: {
    packages: ReferenceSet
    packed: ReferenceSet
  }
  statusSplit: LabelValue[]
  byCategory: LabelValue[]
  byBusinessType: LabelValue[]
  byCustomer: LabelValue[]
  statuses: string[]
  works: string[]
  productCategories: string[]
  businessTypes: string[]
  customers: string[]
}

export async function getPackingDashboard(filters: PackingFilters = {}): Promise<PackingResponse> {
  const { data } = await apiFetch<{ data: {
    kpis: PackingKpis
    status_split: LabelValue[]; by_category: LabelValue[]
    by_business_type: LabelValue[]; by_customer: LabelValue[]
    statuses: string[]; works: string[]; product_categories: string[]
    business_types: string[]; customers: string[]
    references: PackingResponse['references']
    period: ResolvedPeriod
    coverage: Coverage
    data_notes: DataNote[]
    date_field: string
    date_field_options: { value: string; label: string }[]
  } }>(`/dashboard/logistics/packing${buildQuery(filters)}`)
  return {
    kpis: data.kpis,
    statusSplit: data.status_split,
    byCategory: data.by_category,
    byBusinessType: data.by_business_type,
    byCustomer: data.by_customer,
    statuses: data.statuses,
    works: data.works,
    productCategories: data.product_categories,
    businessTypes: data.business_types,
    customers: data.customers,
    references: data.references,
    period: data.period,
    coverage: data.coverage,
    dataNotes: data.data_notes,
    dateField: data.date_field,
    dateFieldOptions: data.date_field_options,
  }
}

// ------------------------------------------------------------------ transport

export interface TransportKpis {
  jobs_shown: number
  delivered: number
  in_progress: number
  total_freight: number
  total_savings: number
}

export interface TransportFilters {
  status?: string[]
  movement_type?: string[]
  source?: string[]
  payment_status?: string[]
  transporter?: string[]
  customer?: string[]
  province?: string[]
  exec_from?: string
  exec_to?: string
  search?: string
  date_from?: string
  date_to?: string
  /** etd (execution) | eta (arrival at works) */
  date_field?: string
}

export interface TransportResponse extends TabMeta {
  kpis: TransportKpis
  references: {
    jobs: ReferenceSet
    delivered: ReferenceSet
    in_progress: ReferenceSet
  }
  statusSplit: LabelValue[]
  byMovementType: LabelValue[]
  byTransporter: LabelValue[]
  byPaymentStatus: LabelValue[]
  byCustomer: LabelValue[]
  byProvince: LabelValue[]
  statuses: string[]
  /** Includes "Unclassified" — 207 jobs genuinely state no movement type. */
  movementTypes: string[]
  sources: string[]
  paymentStatuses: string[]
  transporters: string[]
  customers: string[]
  provinces: string[]
}

export async function getTransportDashboard(filters: TransportFilters = {}): Promise<TransportResponse> {
  const { data } = await apiFetch<{ data: {
    kpis: TransportKpis
    status_split: LabelValue[]; by_movement_type: LabelValue[]
    by_transporter: LabelValue[]; by_payment_status: LabelValue[]
    by_customer: LabelValue[]; by_province: LabelValue[]
    statuses: string[]; movement_types: string[]; sources: string[]
    payment_statuses: string[]; transporters: string[]
    customers: string[]; provinces: string[]
    references: TransportResponse['references']
    period: ResolvedPeriod
    coverage: Coverage
    data_notes: DataNote[]
    date_field: string
    date_field_options: { value: string; label: string }[]
  } }>(`/dashboard/logistics/transport${buildQuery(filters)}`)
  return {
    kpis: data.kpis,
    statusSplit: data.status_split,
    byMovementType: data.by_movement_type,
    byTransporter: data.by_transporter,
    byPaymentStatus: data.by_payment_status,
    byCustomer: data.by_customer,
    byProvince: data.by_province,
    statuses: data.statuses,
    movementTypes: data.movement_types,
    sources: data.sources,
    paymentStatuses: data.payment_statuses,
    transporters: data.transporters,
    customers: data.customers,
    provinces: data.provinces,
    references: data.references,
    period: data.period,
    coverage: data.coverage,
    dataNotes: data.data_notes,
    dateField: data.date_field,
    dateFieldOptions: data.date_field_options,
  }
}

File: app/dashboard/logistics/routes/logistics_dashboard.py
from app.dashboard.logistics.routes.router import router
from fastapi import Request, HTTPException, Query
from app.database import SessionLocal
from app.auth.authenticate_user import authenticate
from app.auth.authorize_user import authorize
from app.dashboard.logistics.helpers import (
    fetch_orders, fetch_filtered_orders,
    fetch_packages, fetch_filtered_packages,
    fetch_trucking, fetch_filtered_trucking, logistics_links,
)
from app.dashboard.logistics.serializers import (
    serialize_shipments, serialize_packing, serialize_transport,
)
from app.dashboard.logistics.calculations import (
    shipment_stage, transport_status, SHIPMENT_STAGES, TRANSPORT_STATUSES,
    job_customer, job_province,
)
from app.accounts.permissions import CAN_VIEW_LOGISTICS_DASHBOARD
from typing import Optional

from app.dashboard.period import resolve_period, serialize_period
from app.dashboard.data_quality import coverage_note, collect
from app.dashboard.logistics.helpers import (
    shipments_coverage, packing_coverage, transport_coverage, order_type_counts,
    fetch_undated_orders,
    SHIPMENT_DATE_OPTIONS, SHIPMENT_DATE_DEFAULT,
    PACKING_DATE_OPTIONS, PACKING_DATE_DEFAULT,
    TRANSPORT_DATE_OPTIONS, TRANSPORT_DATE_DEFAULT,
    NOT_STATED, UNCLASSIFIED,
)
from datetime import date


#=====================================================
# SHIPMENTS  — GET /dashboard/logistics/shipments
#=====================================================

@router.get("/shipments")
def shipments_dashboard(
    request : Request,
    status : Optional[list[str]] = Query(None),
    stage : Optional[list[str]] = Query(None),
    shipping_line : Optional[list[str]] = Query(None),
    country : Optional[list[str]] = Query(None),
    customer : Optional[list[str]] = Query(None),
    etd_from : Optional[date] = None,
    etd_to : Optional[date] = None,
    search : Optional[str] = None,
    # The dashboard-wide window. Both omitted -> the current month.
    date_from : Optional[date] = None,
    date_to : Optional[date] = None,
    # Which date it applies to: etd (sailing) | eta (arrival).
    date_field : Optional[str] = None,
    ):

    db = SessionLocal()
    try:
        authorize(authenticate(request), CAN_VIEW_LOGISTICS_DASHBOARD, db)

        all_orders = fetch_orders(db)
        period_from, period_to, period_kind = resolve_period(date_from, date_to)

        orders = fetch_filtered_orders(
            db, status, shipping_line, country, customer, etd_from, etd_to, search,
            period_from, period_to, date_field,
        )

        cover = shipments_coverage(db, period_from, period_to, date_field)
        stated = sum(1 for o in orders if o.order_type)
        notes = collect(
            coverage_note(
                cover["rows_in_period"], cover["rows_total"], "logistics orders",
                cover["date_field"],
                "Orders with no date in that column fall in no period at all.",
            ),
            coverage_note(
                stated, len(orders), "orders in this period",
                "a local/export type",
                "The Orders tile counts the split over the whole book, because "
                "local orders carry no date and so fall in no period at all.",
            ),
        )

        # Stage is a derived roll-up of the status, so it is filtered here.
        if stage:
            wanted = set(stage)
            orders = [o for o in orders if shipment_stage(o) in wanted]

        statuses = set()
        shipping_lines = set()
        countries = set()
        customers = set()
        for o in all_orders:
            if o.current_status:
                statuses.add(o.current_status)
            if o.shipping_line:
                shipping_lines.add(o.shipping_line)
            if o.origin_country:
                countries.add(o.origin_country)
            if o.customer_name:
                customers.add(o.customer_name)

        data = {
            **serialize_shipments(
                orders, cover, notes,
                # Undated orders sit outside every window, so they cannot come
                # out of the filtered list and are fetched on their own.
                undated=fetch_undated_orders(db, date_field),
            ),
            "period": serialize_period(period_from, period_to, period_kind),
            "date_field": date_field or SHIPMENT_DATE_DEFAULT,
            "date_field_options": SHIPMENT_DATE_OPTIONS,
            # Export against local, counted IN THE WINDOW, with the orders no
            # window can reach reported alongside — see order_type_counts.
            "order_type_counts": order_type_counts(
                db, period_from, period_to, date_field),
            "statuses": sorted(statuses),
            "stages": SHIPMENT_STAGES,
            "shipping_lines": sorted(shipping_lines),
            "countries": sorted(countries),
            "customers": sorted(customers),
        }
        return {"status_code": 200, "detail": "Shipments dashboard fetched", "data": data}

    except HTTPException:
        db.rollback()
        raise
    except Exception as e:
        print(e)
        db.rollback()
        raise HTTPException(status_code=500, detail="Internal server error")
    finally:
        db.close()


#=====================================================
# PACKING  — GET /dashboard/logistics/packing
#=====================================================

@router.get("/packing")
def packing_dashboard(
    request : Request,
    status : Optional[list[str]] = Query(None),
    works : Optional[list[str]] = Query(None),
    product_category : Optional[list[str]] = Query(None),
    business_type : Optional[list[str]] = Query(None),
    customer : Optional[list[str]] = Query(None),
    packing_from : Optional[date] = None,
    packing_to : Optional[date] = None,
    search : Optional[str] = None,
    date_from : Optional[date] = None,
    date_to : Optional[date] = None,
    # packed | rfd
    date_field : Optional[str] = None,
    ):

    db = SessionLocal()
    try:
        authorize(authenticate(request), CAN_VIEW_LOGISTICS_DASHBOARD, db)

        all_packages = fetch_packages(db)
        period_from, period_to, period_kind = resolve_period(date_from, date_to)

        packages = fetch_filtered_packages(
            db, status, works, product_category, business_type, customer,
            packing_from, packing_to, search,
            period_from, period_to, date_field,
        )

        cover = packing_coverage(db, period_from, period_to, date_field)
        costed = sum(1 for p in packages if p.actual_packing_cost)
        notes = collect(
            coverage_note(
                cover["rows_in_period"], cover["rows_total"], "packages",
                cover["date_field"],
            ),
            coverage_note(
                costed, len(packages), "packages in this period",
                "an actual packing cost",
                "Packing savings cannot be computed without it, so it is "
                "reported as unavailable rather than as zero.",
            ),
        )

        statuses = set()
        works_list = set()
        categories = set()
        business_types = set()
        customers = set()
        for p in all_packages:
            if p.status:
                statuses.add(p.status)
            if p.packing_works:
                works_list.add(p.packing_works)
            order = p.consignment
            if order:
                if order.department:
                    categories.add(order.department)
                if order.order_type:
                    business_types.add(order.order_type)
                if order.customer_name:
                    customers.add(order.customer_name)

        data = {
            **serialize_packing(packages, cover, notes),
            "period": serialize_period(period_from, period_to, period_kind),
            "date_field": date_field or PACKING_DATE_DEFAULT,
            "date_field_options": PACKING_DATE_OPTIONS,
            "statuses": sorted(statuses),
            "works": sorted(works_list),
            "product_categories": sorted(categories),
            "business_types": sorted(business_types),
            "customers": sorted(customers),
        }
        return {"status_code": 200, "detail": "Packing dashboard fetched", "data": data}

    except HTTPException:
        db.rollback()
        raise
    except Exception as e:
        print(e)
        db.rollback()
        raise HTTPException(status_code=500, detail="Internal server error")
    finally:
        db.close()


#=====================================================
# TRANSPORT  — GET /dashboard/logistics/transport
#=====================================================

@router.get("/transport")
def transport_dashboard(
    request : Request,
    status : Optional[list[str]] = Query(None),
    movement_type : Optional[list[str]] = Query(None),
    source : Optional[list[str]] = Query(None),
    payment_status : Optional[list[str]] = Query(None),
    transporter : Optional[list[str]] = Query(None),
    customer : Optional[list[str]] = Query(None),
    province : Optional[list[str]] = Query(None),
    exec_from : Optional[date] = None,
    exec_to : Optional[date] = None,
    search : Optional[str] = None,
    date_from : Optional[date] = None,
    date_to : Optional[date] = None,
    # etd (execution) | eta (arrival at works)
    date_field : Optional[str] = None,
    ):

    db = SessionLocal()
    try:
        authorize(authenticate(request), CAN_VIEW_LOGISTICS_DASHBOARD, db)

        all_jobs = fetch_trucking(db)
        period_from, period_to, period_kind = resolve_period(date_from, date_to)

        jobs = fetch_filtered_trucking(
            db, movement_type, source, payment_status, transporter,
            exec_from, exec_to, search,
            period_from, period_to, date_field,
        )

        cover = transport_coverage(db, period_from, period_to, date_field)
        typed = sum(1 for j in jobs if j.movement_type)
        freighted = sum(1 for j in jobs if j.actual_freight is not None)
        notes = collect(
            coverage_note(
                cover["rows_in_period"], cover["rows_total"], "trucking jobs",
                cover["date_field"],
            ),
            coverage_note(
                typed, len(jobs), "jobs in this period", "a movement type",
                "Those sit in the Unclassified bucket rather than being folded "
                "into Inbound or Outbound, which cannot be inferred.",
            ),
            coverage_note(
                freighted, len(jobs), "jobs in this period", "a freight figure",
                "Cost totals cover only the jobs that recorded one.",
            ),
        )

        # customer / city / province are resolved from the linked logistics
        # order (a local logistics consignment moved to trucking carries them).
        links = logistics_links(db, all_jobs)

        # Transport status is a roll-up over the vehicles, so it is filtered here.
        if status:
            wanted = set(status)
            jobs = [j for j in jobs if transport_status(j) in wanted]
        if customer:
            wanted = set(customer)
            jobs = [j for j in jobs if job_customer(j, links) in wanted]
        if province:
            wanted = set(province)
            jobs = [j for j in jobs if job_province(j, links) in wanted]

        movement_types = set()
        sources = set()
        payment_statuses = set()
        transporters = set()
        customers = set()
        provinces = set()
        for j in all_jobs:
            if j.movement_type:
                movement_types.add(j.movement_type)
            if j.source:
                sources.add(j.source)
            if j.payment_status:
                payment_statuses.add(j.payment_status)
            if j.transporter_name:
                transporters.add(j.transporter_name)
            c = job_customer(j, links)
            if c:
                customers.add(c)
            p = job_province(j, links)
            if p:
                provinces.add(p)

        data = {
            **serialize_transport(jobs, links, cover, notes),
            "period": serialize_period(period_from, period_to, period_kind),
            "date_field": date_field or TRANSPORT_DATE_DEFAULT,
            "date_field_options": TRANSPORT_DATE_OPTIONS,
            # Offered as a selectable bucket: 207 jobs genuinely say nothing,
            # and there is no way to tell which way they went.

            "statuses": TRANSPORT_STATUSES,
            # The values present, PLUS the bucket for the 207 jobs that state
            # no movement type — a real answer, and selectable as one.
            "movement_types": sorted(movement_types) + [UNCLASSIFIED],
            "sources": sorted(sources),
            "payment_statuses": sorted(payment_statuses),
            "transporters": sorted(transporters),
            "customers": sorted(customers),
            "provinces": sorted(provinces),
        }
        return {"status_code": 200, "detail": "Transport dashboard fetched", "data": data}

    except HTTPException:
        db.rollback()
        raise
    except Exception as e:
        print(e)
        db.rollback()
        raise HTTPException(status_code=500, detail="Internal server error")
    finally:
        db.close()
